src/librustc_interface/util.rs

```diff
@@ -526,63 +526,75 @@ pub(crate) fn compute_crate_disambiguator(session: &Session) -> CrateDisambiguat
     CrateDisambiguator::from(hasher.finish::<Fingerprint>())
 }
 
+pub(crate) fn check_attr_crate_type(attrs: &[ast::Attribute], lint_buffer: &mut lint::LintBuffer) {
+    // Unconditionally collect crate types from attributes to make them used
+    for a in attrs.iter() {
+        if a.check_name(sym::crate_type) {
+            if let Some(n) = a.value_str() {
+                let crate_types = vec![
+                    sym::rlib,
+                    sym::dylib,
+                    sym::cdylib,
+                    sym::lib,
+                    sym::staticlib,
+                    sym::proc_dash_macro,
+                    sym::bin
+                ];
+
+                if crate_types.contains(&n) {
```

Comment:
If we did the refactoring I just suggested, I'm imagining this code would be like

```rust
if let Some(_) = self.categorize_crate_type(n) {
    return;
}
```